src/Select.tsx

```diff
@@ -1,103 +1,243 @@
-import React, {useEffect, useState, useMemo, useRef} from "react";
-import {useOutsideClick} from './hooks'
+import React, {
+  useEffect,
+  useState,
+  useMemo,
+  useRef,
+  useCallback,
+} from "react";
+import { useOutsideClick } from "./hooks";
+import { SelectOption } from "./components/SelectOption";
+import { SelectChipWrapper } from "./components/SelectChipWrapper";
+import { SelectChip } from "./components/SelectChip";
 
-export interface Option {
-    value: string | number,
-    label: string,
-}
+type SelectProps<T> = {
+  options: T[];
+  search?: boolean;
+  getLabel?: (value: T | null) => string;
+  getValue?: (value: T | null) => string | number;
+} & (
+  | {
+      multiple: true;
+      value: T[];
+      onChange: (value: T[]) => void;
+    }
+  | {
+      multiple: false;
+      value: T | null;
+      onChange: (value: T | null) => void;
+    }
+);
 
-interface SelectProps<T> {
-    value: T | null
-    onChange: (value: T | null) => void
-    options: T[]
-    search?: boolean;
-    getLabel?: (value: T | null) => string
-}
+const toLowerCaseAndTrim = (value: string): string => {
+  return value.toLowerCase().trim();
+};
+export const Select = <V, T = V extends V[] ? V[number] : V>(
+  props: SelectProps<T>,
+) => {
+  const { options, search, getLabel, getValue } = props;
+  const ref = useRef<HTMLDivElement>(null);
+  const openDropdownToggle = () => setOpen(!open);
```

Comment:
open, лучше брать из коллбэка
```suggestion
  const openDropdownToggle = () => setOpen((open) => !open);
```